File: ui/visual-debugger/src/main/java/org/visual/debugger/Debugger.java
package org.visual.debugger;

import javafx.scene.Scene;
import javafx.stage.WindowEvent;
import org.visual.component.display.VisualStage;
import org.visual.component.util.FxUtil;
import org.visual.debugger.constant.FXMLKey;
import org.visual.debugger.context.AttachSceneContext;
import org.visual.debugger.context.DebuggerContext;

public class Debugger {
  private final Scene rootScene = new Scene(DebuggerContext.load(FXMLKey.LAYOUT));
  private final VisualStage stage = new VisualStage();

  public Debugger() {
    AttachSceneContext.getStage()
        .addListener(
            (a, b, newValue) ->
                newValue.addEventFilter(WindowEvent.WINDOW_CLOSE_REQUEST, c -> stage.close()));
  }

  public void showDebugger() {
    FxUtil.runOnFx(
        () -> {
          stage.setScene(rootScene);
          stage.showAndFocus();
        });
  }
}

File: ui/visual-debugger/src/main/kotlin/org/visual/debugger/Debugger.kt
package org.visual.debugger

import javafx.scene.Scene
import org.visual.component.display.VisualStage
import org.visual.component.util.runOnFX
import org.visual.debugger.constant.FXMLKey
import org.visual.debugger.context.DebuggerContext.load

class Debugger(
    private val sceneId: Long
) {
    private val rootScene by lazy { Scene(load(FXMLKey.LAYOUT)) }

    private val stage by lazy {
        VisualStage()
            .apply {
                userData = sceneId
            }
    }

    fun showDebugger(){
        runOnFX {
            stage.scene = rootScene
            stage.showAndFocus()
        }
    }
}